File: solutions/java/src/main/java/playground/algorithm/MergeTwoSortedLists.java
package playground.algorithm;

import playground.algorithm.common.ListNode;

public class MergeTwoSortedLists {

    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        ListNode mergedRoot = new ListNode(0);
        ListNode it = mergedRoot;
        while (l1 != null && l2 != null) {
            if (l1.val <= l2.val) {
                it.next = new ListNode(l1.val);
                l1 = l1.next;
            } else {
                it.next = new ListNode(l2.val);
                l2 = l2.next;
            }
            it = it.next;
        }

        while (l1 != null) {
            it.next = new ListNode(l1.val);
            l1 = l1.next;
            it = it.next;
        }

        while (l2 != null) {
            it.next = new ListNode(l2.val);
            l2 = l2.next;
            it = it.next;
        }

        return mergedRoot.next;
    }
}

File: solutions/go/src/playground/algorithm/merge_two_sorted_lists.go
package algorithm

func mergeTwoLists(l1 *ListNode, l2 *ListNode) *ListNode {
	root := &ListNode{}
	it := root

	for l1 != nil && l2 != nil {
		if l1.Val <= l2.Val {
			it.Next = &ListNode{Val: l1.Val}
			l1 = l1.Next
		} else {
			it.Next = &ListNode{Val: l2.Val}
			l2 = l2.Next
		}
		it = it.Next
	}

	for l1 != nil {
		it.Next = &ListNode{Val: l1.Val}
		l1 = l1.Next
		it = it.Next
	}

	for l2 != nil {
		it.Next = &ListNode{Val: l2.Val}
		l2 = l2.Next
		it = it.Next
	}

	return root.Next
}
